src/explore-education-statistics-admin/src/components/RouteLeavingGuard.tsx

```diff
@@ -0,0 +1,72 @@
+import ModalConfirm from '@common/components/ModalConfirm';
+import useToggle from '@common/hooks/useToggle';
+import React, { useEffect, useState } from 'react';
+import { Prompt, useLocation } from 'react-router';
+import { useHistory } from 'react-router-dom';
+
+interface Props {
+  blockRouteChange: boolean;
+}
+
+const RouteLeavingGuard = ({ blockRouteChange = false }: Props) => {
+  const location = useLocation();
+  const history = useHistory();
+  const [lastLocation, setLastLocation] = useState(location);
+  const [showModal, toggleShowModal] = useToggle(false);
+  const [confirmedNavigation, toggleConfirmedNavigation] = useToggle(false);
+
+  // Block non-react routes
+  useEffect(() => {
+    const handleBeforeUnload = (event: BeforeUnloadEvent) => {
+      if (blockRouteChange) {
+        event.preventDefault();
+        // eslint-disable-next-line no-param-reassign
+        event.returnValue = '';
+      }
+    };
+
+    window.addEventListener('beforeunload', handleBeforeUnload);
+
+    return () => window.removeEventListener('beforeunload', handleBeforeUnload);
+  }, [blockRouteChange]);
+
+  // Block react routes
+  useEffect(() => {
+    if (confirmedNavigation && lastLocation) {
+      toggleConfirmedNavigation.off();
+      history.push(lastLocation.pathname);
+    }
```

Comment:
Does this need to be in an effect?  Couldn't we put directly inside of the modal's `onConfirm`?

Reply:
I've tried putting it in the onConfirm, but `history.push(lastLocation.pathname)` doesn't seem to work from there. I can't figure out why, it has the lastLocation ok. So I've left it in the effect.

Reply:
Cool, after reading it again, I think it's because we need to wait for `confirmedNavigation` to have toggled to true to prevent the Prompt's `message` from re-blocking the route when we call `history.push`!